core-primitives/stf-primitives/src/traits.rs

```diff
@@ -47,3 +49,26 @@ pub trait TrustedCallVerification {
 pub trait PoolTransactionValidation {
 	fn validate(&self) -> Result<ValidTransaction, TransactionValidityError>;
 }
+
+/// Trait to be implemented on the executor to serve helper methods of the executor
+/// to the `IndirectDispatch` implementation.
+pub trait IndirectExecutor<TCS, Error>
+where
+	TCS: PartialEq + Encode + Decode + Debug + Clone + Send + Sync + TrustedCallVerification,
+{
+	fn submit_trusted_call(&self, shard: ShardIdentifier, encrypted_trusted_call: Vec<u8>);
+
+	fn decrypt(&self, encrypted: &[u8]) -> Result<Vec<u8>, Error>;
+
+	fn encrypt(&self, value: &[u8]) -> Result<Vec<u8>, Error>;
+
+	fn get_enclave_account(&self) -> Result<AccountId, Error>;
+
+	fn get_default_shard(&self) -> ShardIdentifier;
+
+	fn sign_call_with_self<TC: Encode + Debug + TrustedCallSigning<TCS>>(
+		&self,
+		trusted_call: &TC,
+		shard: &ShardIdentifier,
+	) -> Result<TCS, Error>;
+}
```

Comment:
What was your reason to introduce those generic types here? Did you make purposely favour it to associated types? In my opinion, associated types are in edge cases a bit more flexible, as you don't have to add them as phantom types to the struct you want to implement a trait on.

This is more of an intuition but I would declare the type trait like this:


```rust
pub trait IndirectExecutor
where
{
      type TC;
      
      type: Error

	fn submit_trusted_call(&self, shard: ShardIdentifier, encrypted_trusted_call: Vec<u8>);

	fn decrypt(&self, encrypted: &[u8]) -> Result<Vec<u8>, Error>;

	fn encrypt(&self, value: &[u8]) -> Result<Vec<u8>, Error>;

	fn get_enclave_account(&self) -> Result<AccountId, Error>;

	fn get_default_shard(&self) -> ShardIdentifier;

	fn sign_call_with_self<TC: Encode + Debug + TrustedCallSigning<TCS>>(
		&self,
		trusted_call: &TC,
		shard: &ShardIdentifier,
	) -> Result<TCS, Error>;
}

```

Reply:
I agree that it would be cleaner. tried with TrustedCall (see below). No time for this rabbit hole....